### PR TITLE
feat: add HTTP requests

### PR DESCRIPTION
**Pull request by r-o-b-o-t-o**
## Changes Proposed:
- Add the `HttpManager` class. Adds non-blocking requests via `HttpManager::HttpRequest`. Available HTTP verbs are GET, POST, PUT, PATCH, DELETE, OPTIONS. You can optionally pass request headers and a body.
- Uses https://github.com/yhirose/cpp-httplib

## Issues Addressed:
See https://github.com/ElunaLuaEngine/Eluna/pull/382#issuecomment-979380019

## Tests Performed:

- Compiled on Windows 10 x64

## How to Test the Changes:

Here are some C++ examples:
1. Basic GET example:
```cpp
HttpManager::GetInstance()->HttpRequest(HttpMethod::Get, "https://random-word-api.herokuapp.com/word", [](int status, const std::string& body, const httplib::Headers& headers) {
    LOG_INFO("server", "Random word: %s", body);
});
```
2. POST example:
```cpp
HttpManager::GetInstance()->HttpRequest(HttpMethod::Post, "https://jsonplaceholder.typicode.com/posts",
    "{\"userId\": 1, \"title\": \"Foo\", \"body\": \"Bar!\"}", // <-- request body
    "application/json", // <-- content type
    [](int status, const std::string& body, const httplib::Headers& headers) {
        LOG_INFO("server", "Result: %s", body);
    }
);
```
3. Example with request headers:
```cpp
httplib::Headers headers = {
    { "Accept", "application/json" },
    { "User-Agent", "AzerothCore" },
};
HttpManager::GetInstance()->HttpRequest(HttpMethod::Get, "https://postman-echo.com/headers", headers, [](int status, const std::string& body, const httplib::Headers& resHeaders) {
    LOG_INFO("server", "Result: %s", body);
});
```

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/shared/Network/HttpManager.cpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
+#include <thread>
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:1]: Need add copyright info

File: src/server/shared/Network/HttpManager.cpp [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+#include <Logging/Log.h>
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:2]:
```suggestion
#include "Log.h"
```

File: src/server/shared/Network/HttpManager.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
+#ifndef __HTTP_MANAGER_H
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.h:1]: Need add copyright info

File: src/server/worldserver/Main.cpp [hunk 4]
```diff
@@ -274,6 +275,8 @@ int main(int argc, char** argv)
     // Set process priority according to configuration settings
     SetProcessPriority("server.worldserver", sConfigMgr->GetOption<int32>(CONFIG_PROCESSOR_AFFINITY, 0), sConfigMgr->GetOption<bool>(CONFIG_HIGH_PRIORITY, false));
 
+    HttpManager httpManager;
```
Comment (Winfidonarleyan) [inline src/server/worldserver/Main.cpp:278]: Very interesting

File: src/server/shared/Network/HttpManager.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+public:
+    HttpResponse(int statusCode, const std::string& body, const httplib::Headers& headers, HttpCallback cb);
+
+    int statusCode;
+    std::string body;
+    httplib::Headers headers;
+    HttpCallback callback;
+};
+
+
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.h:46]:
```suggestion

```

File: src/server/shared/Network/HttpManager.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+{
+public:
+    HttpResponse(int statusCode, const std::string& body, const httplib::Headers& headers, HttpCallback cb);
+
+    int statusCode;
+    std::string body;
+    httplib::Headers headers;
+    HttpCallback callback;
+};
+class HttpManager
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.h:45]:
```suggestion

class HttpManager
```

File: src/server/shared/Network/HttpManager.cpp [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    condVar.notify_one();
+    workerThread.join();
+    ClearQueues();
+    startedWorkerThread = false;
+}
+
+HttpManager* HttpManager::GetInstance()
+{
+    return HttpManager::instance;
+}
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:122]:
```suggestion
HttpManager* HttpManager::GetInstance()
{
    static HttpManager instance;
    return &instance;
}
```

File: src/server/shared/Network/HttpManager.h [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    MPSCQueue<HttpWorkItem> workQueue;
+    MPSCQueue<HttpResponse> responseQueue;
+    std::thread workerThread;
+    bool startedWorkerThread;
+    std::atomic_bool cancelationToken;
+    std::condition_variable condVar;
+    std::mutex condVarMutex;
+    std::regex parseUrlRegex;
+
+    static HttpManager* instance;
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.h:79]:
```suggestion
```

File: deps/httplib/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+CollectIncludeDirectories(
+  ${CMAKE_CURRENT_SOURCE_DIR}
+  PUBLIC_INCLUDES)
+
+target_include_directories(httplib
+  PUBLIC
+    ${PUBLIC_INCLUDES}
+  PRIVATE
+    ${CMAKE_CURRENT_BINARY_DIR})
```
Comment (Winfidonarleyan) [inline deps/httplib/CMakeLists.txt:23]:
```suggestion
target_include_directories(httplib
  INTERFACE
    ${PUBLIC_INCLUDES}})
```

**Comment by anzz1** [comment]
I wrapped this up in a Eluna module https://github.com/anzz1/eluna-module-httpmanager so it can be loaded on-demand

**Review by Nyeriah — APPROVED** [review]

File: deps/PackageList.txt [hunk 10]
```diff
@@ -58,4 +58,10 @@ recastnavigation (Recast is state of the art navigation mesh construction toolse
 
 {fmt} is an open-source formatting library providing a fast and safe alternative to C stdio and C++ iostreams.
   https://github.com/fmtlib/fmt
-  Version: 7.1.3
\ No newline at end of file
+  Version: 7.1.3
```
Comment (Winfidonarleyan) [inline deps/PackageList.txt:61]:
```suggestion
  Version: 8.1.1
```

File: src/server/shared/Network/HttpManager.cpp [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        {
+            std::unique_lock<std::mutex> lock(condVarMutex);
+            condVar.wait(lock, [&] { return workQueue.Peek() != nullptr || cancelationToken.load(); });
+        }
+
+        if (cancelationToken.load())
+        {
+            break;
+        }
+        if (!workQueue.Peek())
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:152]:
```suggestion

        if (!workQueue.Peek())
```

File: src/server/shared/Network/HttpManager.cpp [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        }
+
+        try
+        {
+            std::string host;
+            std::string path;
+
+            if (!ParseUrl(req->url, host, path))
+            {
+                LOG_ERROR("server", "[HttpManager] Could not parse URL %s", req->url.c_str());
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:171]:
```suggestion
                LOG_ERROR("server", "[HttpManager] Could not parse URL {}", req->url.c_str());
```

File: src/server/shared/Network/HttpManager.cpp [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+            }
+
+            httplib::Client cli(host);
+            SetupClient(&cli);
+
+            httplib::Result res = DoRequest(cli, req, path);
+            httplib::Error err = res.error();
+            if (err != httplib::Error::Success)
+            {
+                LOG_ERROR("server", "[HttpManager] HTTP request error: %s", httplib::to_string(err).c_str());
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:182]:
```suggestion
                LOG_ERROR("server", "[HttpManager] HTTP request error: {}", httplib::to_string(err));
```

File: src/server/shared/Network/HttpManager.cpp [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+                httplib::Client cli2(host);
+                SetupClient(&cli2);
+                res = DoRequest(cli2, req, path);
+            }
+
+            responseQueue.Enqueue(new HttpResponse(res->status, res->body, res->headers, req->callback));
+        }
+        catch (const std::exception& ex)
+        {
+            LOG_ERROR("server", "[HttpManager] HTTP request error: %s", ex.what());
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:206]:
```suggestion
            LOG_ERROR("server", "[HttpManager] HTTP request error: {}", ex.what());
```

File: src/server/shared/Network/HttpManager.cpp [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+    case HttpMethod::Put:
+        return client.Put(path, req->headers, req->body, req->contentType.c_str());
+    case HttpMethod::Patch:
+        return client.Patch(path, req->headers, req->body, req->contentType.c_str());
+    case HttpMethod::Delete:
+        return client.Delete(path, req->headers);
+    case HttpMethod::Options:
+        return client.Options(path, req->headers);
+    default:
+        LOG_ERROR("server", "[HttpManager] HTTP request error: invalid HTTP method %d", req->method);
```
Comment (Winfidonarleyan) [inline src/server/shared/Network/HttpManager.cpp:233]:
```suggestion
        LOG_ERROR("server", "[HttpManager] HTTP request error: invalid HTTP method {}", req->method);
```

**Comment by Kitzunu** [comment]
Status @r-o-b-o-t-o

**Comment by Nyeriah** [comment]
It looks ready to me

**Pull request closed without merging**